### PR TITLE
Heating Coil Logic Overhaul

### PR DESCRIPTION
**Pull request by bartimaeusnek**
- [x] Implement Heat Level Interface
- [x] Add Missing Coils
- [x] Refactor EBF
- [x] Refactor MF
- [x] Refactor Pyro
- [x] Refactor OilCracker

### CODE REVIEW

**Review by leagris — CHANGES_REQUESTED** [review]

File: src/main/java/gregtech/api/enums/GT_Values.java [hunk 1]
```diff
@@ -126,6 +127,16 @@
                     "Ultimate High Voltage", "Ultimate Extreme Voltage", "Ultimate Insane Voltage",
                     "Ultimate Mega Voltage", "Ultimate Extended Mega Voltage", "Overpowered Voltage",
                     "Maximum Voltage"};
+
+    public static final String[] TIER_COLORS =
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/api/enums/GT_Values.java:131]: No. This is an API/Utility Class of which all fields are needed to be public, to be exposed, i could encapsulate the field with getters and setters tho.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 2]
```diff
@@ -34,21 +34,21 @@
...
-    public ArrayList<GT_MetaTileEntity_Hatch_Energy> mEnergyHatches = new ArrayList<GT_MetaTileEntity_Hatch_Energy>();
-    public ArrayList<GT_MetaTileEntity_Hatch_Maintenance> mMaintenanceHatches = new ArrayList<GT_MetaTileEntity_Hatch_Maintenance>();
+    public ArrayList<GT_MetaTileEntity_Hatch_Input> mInputHatches = new ArrayList<>();
+    public ArrayList<GT_MetaTileEntity_Hatch_Output> mOutputHatches = new ArrayList<>();
+    public ArrayList<GT_MetaTileEntity_Hatch_InputBus> mInputBusses = new ArrayList<>();
+    public ArrayList<GT_MetaTileEntity_Hatch_OutputBus> mOutputBusses = new ArrayList<>();
+    public ArrayList<GT_MetaTileEntity_Hatch_Dynamo> mDynamoHatches = new ArrayList<>();
+    public ArrayList<GT_MetaTileEntity_Hatch_Muffler> mMufflerHatches = new ArrayList<>();
+    public ArrayList<GT_MetaTileEntity_Hatch_Energy> mEnergyHatches = new ArrayList<>();
+    public ArrayList<GT_MetaTileEntity_Hatch_Maintenance> mMaintenanceHatches = new ArrayList<>();
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:51]: Ill not change the markup of public API fields, and introduce possible breaking changes for the sake of code smell here

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AbstractMultiFurnace.java [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+
+    public int getPollutionPerTick(ItemStack aStack) {
+        return 20;
+    }
+
+    public int getDamageToComponent(ItemStack aStack) {
+        return 0;
+    }
+
+    public boolean explodesOnComponentBreak(ItemStack aStack) {
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AbstractMultiFurnace.java:89]: aStack is inherited from public api

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AbstractMultiFurnace.java [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+
+    public int getMaxEfficiency(ItemStack aStack) {
+        return 10000;
+    }
+
+    public int getPollutionPerTick(ItemStack aStack) {
+        return 20;
+    }
+
+    public int getDamageToComponent(ItemStack aStack) {
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AbstractMultiFurnace.java:85]: aStack is inherited from public api

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AbstractMultiFurnace.java [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        if (aBaseMetaTileEntity.getBlockOffset(xDir + i, 0, zDir + j) != GregTech_API.sBlockCasings1)
+            return false;
+        return aBaseMetaTileEntity.getMetaIDOffset(xDir + i, 0, zDir + j) == CASING_INDEX;
+    }
+
+    public int getMaxEfficiency(ItemStack aStack) {
+        return 10000;
+    }
+
+    public int getPollutionPerTick(ItemStack aStack) {
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AbstractMultiFurnace.java:81]: aStack is inherited from public api

File: src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AbstractMultiFurnace.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+            return true;
+        if (addEnergyInputToMachineList(tTileEntity, CASING_INDEX))
+            return true;
+
+        if (aBaseMetaTileEntity.getBlockOffset(xDir + i, 0, zDir + j) != GregTech_API.sBlockCasings1)
+            return false;
+        return aBaseMetaTileEntity.getMetaIDOffset(xDir + i, 0, zDir + j) == CASING_INDEX;
+    }
+
+    public int getMaxEfficiency(ItemStack aStack) {
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/common/tileentities/machines/multi/GT_MetaTileEntity_AbstractMultiFurnace.java:77]: aStack is inherited from public api

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 7]
```diff
@@ -1003,4 +1003,45 @@ public boolean allowPullStack(IGregTechTileEntity aBaseMetaTileEntity, int aInde
     public boolean allowPutStack(IGregTechTileEntity aBaseMetaTileEntity, int aIndex, byte aSide, ItemStack aStack) {
         return false;
     }
+
+    protected ItemStack[] getCompactedInputs(){
+        ArrayList<ItemStack> tInputList = getStoredInputs();
+        int tInputList_sS = tInputList.size();
+        for (int i = 0; i < tInputList_sS - 1; i++) {
+            for (int j = i + 1; j < tInputList_sS; j++) {
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:1011]: The use of continue in this case is less smelly than a 3rd nesting level imo.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 8]
```diff
@@ -1003,4 +1003,45 @@ public boolean allowPullStack(IGregTechTileEntity aBaseMetaTileEntity, int aInde
...
+                if (!GT_Utility.areStacksEqual(tInputList.get(i), tInputList.get(j)))
+                    continue;
+                if (tInputList.get(i).stackSize >= tInputList.get(j).stackSize) {
+                    tInputList.remove(j--);
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:1015]: I doubt thats possible in this case. Im open to any solutions tho.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 9]
```diff
@@ -1003,4 +1003,45 @@ public boolean allowPullStack(IGregTechTileEntity aBaseMetaTileEntity, int aInde
...
+                    tInputList_sS = tInputList.size();
+                } else {
+                    tInputList.remove(i--);
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:1018]: I doubt thats possible in this case. Im open to any solutions tho.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 10]
```diff
@@ -1003,4 +1003,45 @@ public boolean allowPullStack(IGregTechTileEntity aBaseMetaTileEntity, int aInde
...
+    protected FluidStack[] getCompactedFluids(){
+        ArrayList<FluidStack> tFluidList = getStoredFluids();
+        int tFluidList_sS = tFluidList.size();
+        for (int i = 0; i < tFluidList_sS - 1; i++) {
+            for (int j = i + 1; j < tFluidList_sS; j++) {
+                if (!GT_Utility.areFluidsEqual(tFluidList.get(i), tFluidList.get(j)))
+                    continue;
+
+                if (tFluidList.get(i).amount >= tFluidList.get(j).amount) {
+                    tFluidList.remove(j--);
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:1036]: I doubt thats possible in this case. Im open to any solutions tho.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 11]
```diff
@@ -1003,4 +1003,45 @@ public boolean allowPullStack(IGregTechTileEntity aBaseMetaTileEntity, int aInde
...
+                    tFluidList_sS = tFluidList.size();
+                } else {
+                    tFluidList.remove(i--);
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:1039]: I doubt thats possible in this case. Im open to any solutions tho.

File: src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java [hunk 12]
```diff
@@ -1003,4 +1003,45 @@ public boolean allowPullStack(IGregTechTileEntity aBaseMetaTileEntity, int aInde
...
+            }
+        }
+        return tInputList.toArray(new ItemStack[0]);
+    }
+
+    protected FluidStack[] getCompactedFluids(){
+        ArrayList<FluidStack> tFluidList = getStoredFluids();
+        int tFluidList_sS = tFluidList.size();
+        for (int i = 0; i < tFluidList_sS - 1; i++) {
+            for (int j = i + 1; j < tFluidList_sS; j++) {
```
[the remark this answers is not in the document]
Comment (bartimaeusnek) [reply src/main/java/gregtech/api/metatileentity/implementations/GT_MetaTileEntity_MultiBlockBase.java:1031]: The use of continue in this case is less smelly than a 3rd nesting level imo.

**Review by leagris — APPROVED** [review]

**Comment by Prometheus0000** [comment]
Maybe you should mention what you changed did, so people don't have to hunt through the code to figure it out?

Like apparently coils speed up oil crackers now for example. How am I supposed to know this without hunting through code? I think people wanted perfect OCs for LCRs based on coils, but this seems like the pyro version, and I don't know which people would want. I know it's a different multi, but did you discuss with anyone in what way it should be changed, or whether it should at all?

**Pull request merged**